src/felzenszwalb.jl

```diff
@@ -0,0 +1,184 @@
+        end
+        return parent[x]
+    end
+
+    function merge_trees(x, y)
+        x = Int(x)
+        y = Int(y)
+        num_sets -= 1
+        if rank[x]<rank[y]
+            parent[x] = y
+            set_size[y] += set_size[x]
+            return y
+        elseif rank[x]>rank[y]
+            parent[y] = x
+            set_size[x] += set_size[y]
+            return x
+        else
+            parent[y] = x
+            set_size[x] += set_size[y]
+            rank[x] += 1
+            return x
+        end
+    end
+
+    sort(edges, lt = (x,y)->(x[3]<y[3]))
+
+    for i in range(1,num_edges)
+        edge = edges[i]
+        a, b, w = edge
+        a = find_root(a)
+        b = find_root(b)
+        if a!=b
+            if w <= min(threshold[a], threshold[b])
+                merged_root = merge_trees(a, b)
+                threshold[merged_root] = w + k/set_size[merged_root]
+            end
+        end
+    end
+
+    segments = OrderedSet()
+    for i in 1:num_vertices
+        push!(segments, find_root(i))
+    end
+
+    segments2index = Dict{Int, Int}()
+    for i in 1:num_sets
+        segments2index[segments[i]]=i 
+    end
+
+    index_map = Array{Int64}(num_vertices)
+    for i in 1:num_vertices
+        index_map[i] = segments2index[find_root(i)]
+    end
+
+    return index_map, num_sets
+end
+
+function felzenszwalb{T<:Gray}(img::AbstractArray{T, 2}, k::Real; sigma=0.8)
+
+    #img = imfilter(img, Kernel.gaussian(sigma))
+    rows, cols = size(img)
+    num_vertices = rows*cols
+    num_edges = 4*rows*cols - 3*rows - 3*cols + 2
+    edges = Array{SVector{3, Float64}}(num_edges)
+
+    R = CartesianRange(size(img))
+    I1, Iend = first(R), last(R)
+    num = 1
+    for I in R
+        for J in CartesianRange(max(I1, I-I1), min(Iend, I+I1))
+            if I >= J
+                continue
+            end
+            edges[num] = SVector((I[2]-1)*rows+I[1], (J[2]-1)*rows+J[1], abs(img[I]-img[J]))
+            num += 1
+        end
+    end
+
+    index_map, num_segments = felzenszwalb_graph(edges, num_vertices, k)
+    result = similar(img, Int64)
+    labels = Array(1:num_segments)
+    region_means = Dict{Int, T}()
+    region_pix_count = Dict{Int, Int}()
+
+    for j in indices(img, 2)
+        for i in indices(img, 1)
+            result[i, j] = index_map[(j-1)*rows+i]
+        end
+    end
+
+    for i in range(1, num_segments)
+        region = find(x->x==i, result)
+        region_pix_count[i] = length(region)
+        region_means[i] = mean(img[region])
+    end
+
+    return SegmentedImage(result, labels, region_means, region_pix_count) 
+
+end
+
+function felzenszwalb{T<:Color}(img::AbstractArray{T, 2}, k::Real; sigma=0.8)
+
+    rows, cols = size(img)
+    num_vertices = rows*cols
+    num_edges = 4*rows*cols - 3*rows - 3*cols + 2
+    edges = Array{SVector{3, Float64}}(num_edges)
+    channel_index_map = Array{Int}(num_vertices, 3)
+    channel_num_segments = Array{Int}(3)
```

Comment:
ColorVectorSpace supports math on RGB values. Could one create a single graph, just using the color-difference? Stated as a question, what advantage does generating 3 different graphs convey?